[task-manager] Make TaskManagerTaskBody data generic

Noitidart opened the pull request:
For instance, this now becomes possible:

```
import { LocationObject } from 'expo-location';
import { TaskManagerTaskBody } from 'expo-task-manager';

function handleLocationChange(input: TaskManagerTaskBody<{ locations: LocationObject[] }>) {

}
```

# Why

<!-- 
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests. 
-->

Improved Typescript

# How
Just types, no build.
<!-- 
How did you build this feature or fix this bug and why? 
-->

# Test Plan

Just types, not needed.
<!-- 
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction. 
-->

tsapeta reviewed (changes requested):
My comment (specifically 1. and 4. points) from another PR also applies here: https://github.com/expo/expo/pull/11670#pullrequestreview-571061541

Noitidart commented:
Thank you @tsapeta for the fast and detailed review. I have rebased a fix.

Noitidart committed:
Add generic to TaskManagerTaskBody

For instance, this now becomes possible:

```
import { LocationObject } from 'expo-location';
import { TaskManagerTaskBody } from 'expo-task-manager';

function handleLocationChange(input: TaskManagerTaskBody<{ locations: LocationObject[] }>) {

}
```
